### ISSUE TITLE
Docker service limit io disk

### ISSUE THREAD

**Issue by cristhianbicca** [docker/docker#32602]
It would be more of a doubt but I do not know where to ask this:

Would there be any way to limit the disk IO of the containers with "docker service"?

Something like this "docker create"?

      --device-read-bps list       Limit read rate (bytes per second) from a device (default [])
      --device-read-iops list      Limit read rate (IO per second) from a device (default [])
      --device-write-bps list      Limit write rate (bytes per second) to a device (default [])
      --device-write-iops list     Limit write rate (IO per second) to a device (default [])

Thanks

**Comment by thaJeztah** [comment]
Thanks for suggesting; I linked this issue from https://github.com/docker/docker/issues/25303, which is the "epic" we're using to track `docker run` options that are not (or not _yet_) implemented for services.

This feature may be difficult to implement exactly the same as the option or `docker run`, given that limiting io is set for a specific _device_; in a swarm situation, it's not possible to predict what node a service's tasks (containers) will be deployed on, and that node may not have the device that's specified in the option

/cc @ehazlett @aluzzardi

**Comment by cristhianbicca** [comment]
Thanks @thaJeztah

**Comment by svanoort** [comment]
@thaJeztah Determining the device is already a bit fiddly unless you know you're always launching in the same environment -- it's more or less necessary to launch a container and then inspect the block devices list.  Even with the challenges, it's an absolute killer feature for containerization.

FWIW I'd love to see a version of this in swarm & compose, even if it *DOESN'T* try to solve the device naming problem.  Currently I'm building a scalability testing platform for one of the bigger open source applications, and using device IOPS/throughput limits to simulate weaker hardware (precursor to major implementation changes).  Even an imperfect version of this feature in compose/swarm would let me simplify greatly.

**Comment by svanoort** [comment]
Alternately, allowing the device name to accept a wildcard of some sort might very well be enough to make this usable across a swarm, such as /dev/sd* to hit hard drives or /dev/?da* similarly to cover (for example) hard drives and virtual drives.   

This would also make it considerably easier to use the device limits command too, because you need not specifically identify the device in question.

**Comment by thaJeztah** [comment]
Yes, I think ideally, this concept would be a bit more "abstracted" to make it more portable (and I like the idea of the wildcards); perhaps even combining the `bps` / `iops`  to a single option (`--some-limit device=/dev/sd*,limit=1234iops`) - can they both be set, and do they make sense to be combined (? - wondering).

While it's not a full solution, perhaps your use-case could be resolved by setting a default parent cgroup on the daemon (https://docs.docker.com/engine/reference/commandline/dockerd/#default-cgroup-parent); this should work even for services, because any container started on such a host would use this option.

**Comment by cirocosta** [comment]
I +1 on allowing wildcards, as I see it would help in many cases. What I've been doing so far is always have a raid configuration at a known device location which is then easy to target. On top of that we forked Runc to place default constraints to groups (not the best solution as I see :grimacing: ) and then used it with the default runtime setting on dockerd.

Hey @thaJeztah , as I understand the default cgroup parent wouldn't work for this use case as it'd make all containers inherit a single cgroup with a given limit (i.e, he'd not limit each container individually but actually the whole group of containers) :thinking: 

For instance:

```sh
 sudo mkdir /sys/fs/cgroup/pids/mygroup
sudo echo "20" | sudo tee /sys/fs/cgroup/pids/mygroup/pids.max
 docker run --cgroup-parent /mygroup -d nginx:alpine
 docker run --cgroup-parent /mygroup -d nginx:alpine

/sys/fs/cgroup/pids/mygroup
├── 80e24e4eeb7a113a714cec0649992a6256432bb17d6008593102b6952c2fb6a9
...
│   ├── pids.max
│   └── tasks
├── b238147829ceab968d6ec0bf4a12319500646902c651acb17552b1d8163bced8
...
│   ├── pids.max
│   └── tasks
...
├── pids.max
└── tasks
```

Is it possible to make cgroup-parent act in a "per-container" container? Such that when I create these other two containers they would not inherit the 20 pids limit and actually each have a 20 pid limit.

---

**Comment by svanoort** [comment]
Yes, @thaJeztah as @cirocosta notes, the cgroup-parent solution does not work in this case -- we're launching a group of containers and only want one to be resource-limited. I suspect this represents a fairly common use-case, where often someone doesn't want to ensure one application doesn't consume excessive IO resources in a cluster. 

Here's what the topology looks like:

* One Jenkins master container (the application), with resource limits applied.  Container placement restrictions will facilitate placing this on a different node from the other containers when possible.
* One instance of a git (providing data to Jenkins and for build agents)
* One instance of Graphite, to collect resource use and performance metrics streamed from Jenkins
* A whole fleet of containers for Jenkins build agents (to ensure the master is only doing orchestration work)
* Fun point: this may be running locally on native Linux Docker  or Mac Docker, or spread across multiple cloud instances

I'm also not entirely sure if there is a solid way to do it, but likely we'll need to do something to ensure we don't accidentally end up using volumes for key I/O (currently I'm assuming resource limits don't apply to them, but cannot find any documentation on this).  

In any case, wildcards for devices + device resource limits in compose/swarm would 90% cover needs, and also cover by far the broadest use case for most users.

**Comment by japrogramer** [comment]
I would Like to see support for this in docker swarm deploy.
https://docs.docker.com/compose/compose-file/compose-file-v2/#blkio_config
